qiskit/optimization/converters/optimization_problem_to_negative_value_oracle.py

```diff
@@ -0,0 +1,230 @@
+        self._verbose = verbose
+        self._backend = backend
+
+    def encode(self, linear_coeff, quadratic_coeff, constant):
+        """
+        A helper function that converts a quadratic function into a state preparation operator A and
+         oracle O that recognizes negative numbers.
+        Args:
+            linear_coeff (np.array): The linear coefficients of the QUBO.
+            quadratic_coeff (np.array): The quadratic coefficients of the QUBO.
+            constant (int): The constant of the QUBO.
+        Returns:
+            Tuple(InitialState.Custom, CustomCircuitOracle, dict): A state preparation operator A
+            (InitialState), an oracle O (CustomCircuitOracle) that recognizes negative numbers, and
+            a dictionary representation of the function coefficients, where the key -1 represents
+            the constant.
+        """
+        # Get circuit requirements from input.
+        num_key = len(linear_coeff)
+        num_ancilla = max(num_key, self._num_value) - 1
+
+        # Get the function dictionary.
+        func = self.__get_function(num_key, linear_coeff, quadratic_coeff, constant)
+        if self._verbose:
+            print("f:", func, "\n")
+
+        # Define state preparation operator A from function.
+        quantum_dict = QQUBODictionary(num_key, self._num_value, num_ancilla,
+                                       func, backend=self._backend)
+        a_operator_circuit = quantum_dict.circuit
+        a_operator = Custom(a_operator_circuit.width(), circuit=a_operator_circuit)
+
+        # Get registers from the A operator circuit.
+        reg_map = {}
+        for reg in a_operator_circuit.qregs:
+            reg_map[reg.name] = reg
+        key_val = reg_map["key_value"]
+        anc = reg_map["ancilla"]
+
+        # Build negative value oracle O.
+        oracle_bit = QuantumRegister(1, "oracle")
+        oracle_circuit = QuantumCircuit(key_val, anc, oracle_bit)
+        self.__cxzxz(oracle_circuit, key_val[num_key], oracle_bit[0])
+        oracle = CustomCircuitOracle(variable_register=key_val,
+                                     output_register=oracle_bit,
+                                     ancillary_register=anc,
+                                     circuit=oracle_circuit,
+                                     evaluate_classically_callback=self.__evaluate_classically)
+
+        return a_operator, oracle, func
+
+    @staticmethod
+    def __get_function(num_assets, linear, quadratic, constant):
+        """Convert the problem to a dictionary format."""
+        func = {-1: constant}
+        for i in range(num_assets):
+            func[i] = -linear[i]
+            for j in range(i):
+                func[(i, j)] = int(quadratic[i, j])
+
+        return func
+
+    @staticmethod
+    def __cxzxz(circuit, ctrl, tgt):
+        """Multiplies by -1."""
+        circuit.cx(ctrl, tgt)
+        circuit.cz(ctrl, tgt)
+        circuit.cx(ctrl, tgt)
+        circuit.cz(ctrl, tgt)
```

Comment:
Two questions:
1) This multiplies by -1 if the control qubit is set. Applying `Z` to the control qubit directly should give the same result, shouldn't it?
2) If just applying  `Z` is not the same, the `CX` gates can be replaced by just `X` gates on the target, since they just cancel each other if the control qubit is 0.

Reply:
I think in this use case it might be the same - I just pulled the generic util from our library. However, I can just remove this if I can figure out how to use the LogicalExpressionOracle instead of CustomCircuitOracle.

Reply:
I can't use LogicalExpressionOracle because I can't set the variable register. But it seems you are right - applying Z to the control is equivalent. Removed CXZXZ and simplified the CustomCircuitOracle setup.